File: java/98.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // public boolean isValid(TreeNode node, Integer lower, Integer upper){
    //     if(node == null) return true;
    //     if(lower != null && node.val <= lower) return false;
    //     if(upper != null && node.val >= upper) return false;
    //     return isValid(node.left, lower, node.val) && isValid(node.right, node.val, upper);
    // }
    // public boolean isValidBST(TreeNode root) {
    //     return isValid(root, null, null);
    // }
    
    public boolean isValidBST(TreeNode root){
        if(root == null) return true;
            Stack<TreeNode> stack = new Stack();
            double pre = -Double.MAX_VALUE;
            while (!stack.empty() || root != null){
                while(root != null){
                    stack.push(root);
                    root = root.left;
                }
                root = stack.pop();
                if(root.val <= pre)
                    return false;
                pre = root.val;
                root = root.right;
            }
            return true;
    }
}

File: python/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # def isValidBST(self, root: TreeNode) -> bool:
    #     def helper(node, lower=float('-inf'), upper=float('inf')):
    #         if not node:
    #             return True
    #         if node.val <= lower or node.val >= upper:
    #             return False
    #         return helper(node.left, lower, node.val) and helper(node.right, node.val, upper)
    #     return helper(root)
    def isValidBST(self, root: TreeNode) -> bool:
        stack, pre = [], float('-inf')
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if root.val <= pre:
                return False
            pre = root.val
            root = root.right
        return True
